CMake/Versioning.cmake

```diff
@@ -0,0 +1,125 @@
+#
+# Versioning
+#
+# Usage:
+#
+#   include(Versioning.cmake)
+#
+# Description:
+#
+# The root folder of the project contains a `VERSION` file.
+# This `VERSION` file contains a version string, e.g. `1.2.3`.
+# This version string is changed by the maintainer manually,
+# before building and tagging the next version for release.
+# By using a standalone versioning file you don't have to touch CMakeLists.txt.
+#
+# The version string is tokenized into [major].[minor].[patch] tokens.
+#
+# Then additional versioning information is collected and added.
+# This allows to add any static or dynamic strings to the version.
+# A dynamic version string might come from the source-control-system, e.g. git.
+# That enables you to add, e.g. the fetched short-hash of a git commit,
+# the git branch name or a current timestamp as release date.
+# You could add the static string, e.g. "dev-local", if is not a git repo.
+#
+# All the version strings are then inserted into the version template file,
+# which uses variable replacement tags (@VAR@).
+#
+# Finally, the new version header file is written.
+#
+
+#-------------------------------------------------------------------
+# Read Version from VERSION file
+#-------------------------------------------------------------------
+
+file(STRINGS "${CMAKE_CURRENT_LIST_DIR}/../VERSION" version)
```

Comment:
Can we change this to use:

```
> git describe --tags --abbrev=40 --dirty --broken --long --match="v[0-9][0-9.]*"
v0.2.3-13-g09d1ced097d9d9c17529275aa7c43dd1a245de62-dirty
```

That way we avoid having maintaining synchronisation between the VERSION file and the git tags.

The result of that describe mains:

- **v0.2.3**: tag name; which matches v[0-9][0-9.]*, we probably should allow beta, and release candidates as well.
- **13**: Number of commits since the tag
- **g09d1ced097d9d9c17529275aa7c43dd1a245de62**: The full git sha
- **dirty**: The are uncommitted changes, this instead may show **broken** when the git repository is corrupt. or nothing at all if freshly checked out.

The only thing I am worried about with this technique is if when releasing with github, if the tag is committed before it does a build.

Reply:
That means you don't want to have a fixed version number in your source and depend on the `git tag`?
In other words: If you switch this to be depending on git for the inital base version (v1.2.3), 
you won't have a version number for the zip/tar of your source.

For instance, my situation on Appveyor: i need a base version (v1.2.3) without having a tag...  
The version string is currently only the continous part, `dev-shorthash`, e.g. `ttauri-dev-a4fa343a`, while it would be nice to have `ttauri-v1.2.3-dev-a4fa343a-msvc2019-x64.zip`, right?

Reply:
In my previous work, we really needed to know if this was a release build, or some kind of other build.

With the commits since tag, dirty and also the branch name we could easily figure out what kind of build this was. Of course if the branch is "main" or if there are 0 commits since the tag, we would omit this from the filename of the executable/zip file.

A short hash would be fine I guess. (the --long has to do with the format of the output string so that we can parse this easy because all the elements are always there). Using --abbrev=0 should get us a short hash value.

So yea the zip file having the short description, with optional commits since tag, and branch name would be nice.

Not sure about the .zip file of the source, does github not create the zip file for the source already based on the tag given when you make a release?

Reply:
Oh, i see. Never needed these additional values for anything. That's why i'm asking.
The only other thing we used a compact version number for is checking for deprecated versions.

Well, it will be a very long filename then :)

> Not sure about the .zip file of the source, does github not create the zip file for the source already based on the tag given when you make a release? 

Yes, it does, e.g you get `v2.3.0.zip`. But when you extract it you have files without a version number and being not in a .git repo.

Reply:
As I said for the filename, use the short format. But with all the optional stuff added to it.